### PR TITLE
Bindeps and AbstractNLPModel

### PR DESCRIPTION
**Pull request by abelsiqueira**
Two major changes.

- bindeps: Install CUTEst locally automatically for use inside CUTEst.jl
- use NLPModels.jl: Last "basic" main package for NLPModels.jl (Along with JuMP, Ampl and Simple)

### CODE REVIEW

**Comment by dpo** [comment]
Thanks! I'm expecting an issue on OSX because brew currently ships gcc and gfortran 6.1, which are buggy and can't compile sifdecode.

**Comment by lrsantos11** [comment]
@dpo there is any way out to install CUTEst.jl on OSX without this gfortran bug for sifdecode? @abelsiqueira pull request #52 seems to make this gfortran independent?

**Comment by dpo** [comment]
@lrsantos11 Try this:
```bash
$ brew update  # I just updated archdefs.rb
$ brew tap homebrew/versions  # for gcc-5 and gfortran-5, which are ok
$ brew install gcc5
$ CUTEST_GCC=gcc-5 CUTEST_GFORTRAN=gfortran-5 brew install archdefs
$ brew install cutest
```

**Comment by dpo** [comment]
@abelsiqueira `.travis.yml` should probably do the same.

File: deps/build.jl [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+end
+
+@osx_only begin
+  try
+    run(`brew doctor`)
+  end
+  run(`brew update`)
+  run(`brew tap optimizers/cutest`)
+  run(`brew install cutest`)
+  run(`brew install mastsif`)
```
Comment (dpo) [inline deps/build.jl]: Would this work with Homebrew.jl?

**Comment by abelsiqueira** [comment]
I just pushed a possible change using `Homebrew.jl` with the new commands to a different branch.
Since I don't use osx, we'll have to wait for Travis.

**Comment by lrsantos11** [comment]
Thanks @dpo. It's installed like a charm.

**Comment by abelsiqueira** [comment]
@dpo Now using Homebrew.jl.
Julia nightly for OSX is not working for some unknown reason. It hangs on the tests, which may indicate they are taking too long. But they weren't before.

By the way, this supercedes #52.

File: deps/build.jl [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+        var = split(split(line, "=")[1])[2]
+        path = chomp(split(line, "=")[2])
+        write(cenv, "ENV[\"$var\"] = \"$path\"\n")
+      end
+    end
+  end
+end
+
+@osx_only begin
+  @eval Pkg.add("Homebrew")
```
Comment (dpo) [inline deps/build.jl]: This should be in the `REQUIRE` file.

Comment (abelsiqueira) [inline deps/build.jl]: But Linux doesn't require it.

Comment (dpo) [inline deps/build.jl]: You can use the `@osx_only` syntax in `REQUIRE`.

**Comment by abelsiqueira** [comment]
Force pushed correction

**Comment by dpo** [comment]
Any idea what this is: https://travis-ci.org/JuliaOptimizers/CUTEst.jl/jobs/140419572#L815 ?

**Comment by abelsiqueira** [comment]
I think there is some random error, maybe because of the force push. My homebrew branch had the same error, and I also force pushed it. But my abstractnlp branch was not force pushed and did not produce error.
Notice the [commits](https://travis-ci.org/abelsiqueira/CUTEst.jl/builds).

I'm gonna ask for a restart in the build. Let's see what happens.

**Comment by abelsiqueira** [comment]
Magically works now.

File: deps/build.jl [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+        var = split(split(line, "=")[1])[2]
+        path = chomp(split(line, "=")[2])
+        write(cenv, "ENV[\"$var\"] = \"$path\"\n")
+      end
+    end
+  end
+end
+
+@osx_only begin
+  @eval using Homebrew
```
Comment (dpo) [inline deps/build.jl]: I don't think there's a need to `@eval` here.

File: deps/build.jl [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+  Homebrew.update()
+  Homebrew.add("homebrew/versions/gcc5")
```
Comment (dpo) [inline deps/build.jl]: Need to `tap homebrew/versions`.

Comment (dpo) [inline deps/build.jl]: Ah, maybe not.

File: deps/build.jl (hunk 3 above)
Comment (abelsiqueira) [inline deps/build.jl]: @eval was due to some error regarding `using` not being on the top level.
There was some explaining that I didn't quite understood, but using @eval
fixed the problem.
Some simple code that may reproduce the error is
````
@osx_only begin
using Plots
for i = 1:3
    println(i)
end
end
````

Comment (dpo) [inline deps/build.jl]: Strange. I do it in `HSL.jl`.

**Comment by dpo** [comment]
Actually, if the user already has CUTEst installed, this reinstalls the entire thing, including gcc-5. It'd be better to detect if `libcutest` is already available or not. See what I did in MUMPS.jl for example.

**Comment by abelsiqueira** [comment]
I'm gonna create a issue to tackle this later.

**Comment by dpo** [comment]
@abelsiqueira @lrsantos11 Homebrew now ships gcc/gfortran 6.2, which don't have the bug above. So there's no need to use gcc-5 anymore.

**Pull request merged**